File: src/chapters/interpolation/content/code/lagrange.cpp
#include <vector>
#include <iostream>
#include <cmath>
using namespace std;
using Vec = vector<double>;
using Mat = vector<Vec>;

// Solve A x = b by Gaussian elimination with partial pivoting.
Vec solve(Mat A, Vec b) {
    int n = b.size();
    for (int k = 0; k < n; ++k) {
        int p = k;
        for (int i = k + 1; i < n; ++i) if (fabs(A[i][k]) > fabs(A[p][k])) p = i;
        swap(A[k], A[p]); swap(b[k], b[p]);
        for (int i = k + 1; i < n; ++i) {
            double f = A[i][k] / A[k][k];
            for (int j = k; j < n; ++j) A[i][j] -= f * A[k][j];
            b[i] -= f * b[k];
        }
    }
    Vec x(n);
    for (int i = n - 1; i >= 0; --i) {
        double s = b[i];
        for (int j = i + 1; j < n; ++j) s -= A[i][j] * x[j];
        x[i] = s / A[i][i];
    }
    return x;
}

// Coefficients (low->high) of the interpolant through (x_i, y_i) via Vandermonde.
Vec lagrange_coeffs(const Vec& x, const Vec& y) {
    int n = x.size();
    Mat V(n, Vec(n, 1.0));
    for (int i = 0; i < n; ++i)
        for (int j = 1; j < n; ++j) V[i][j] = V[i][j - 1] * x[i];   // x_i^j
    return solve(V, y);
}

int main() {
    Vec x = {-1, 1, 2, 3}, y = {-3, 1, 3, 29};
    Vec a = lagrange_coeffs(x, y);
    cout << "coefficients (low->high):";
    for (double v : a) cout << " " << v;
    cout << "\n";
}

File: src/chapters/interpolation/content/code/lagrange.rs
// Lagrange interpolation via the Vandermonde system V a = y.
fn solve(mut a: Vec<Vec<f64>>, mut b: Vec<f64>) -> Vec<f64> {
    let n = b.len();
    for k in 0..n {
        let mut p = k;
        for i in k + 1..n { if a[i][k].abs() > a[p][k].abs() { p = i; } }
        a.swap(k, p); b.swap(k, p);
        for i in k + 1..n {
            let f = a[i][k] / a[k][k];
            for j in k..n { a[i][j] -= f * a[k][j]; }
            b[i] -= f * b[k];
        }
    }
    let mut x = vec![0.0; n];
    for i in (0..n).rev() {
        let mut s = b[i];
        for j in i + 1..n { s -= a[i][j] * x[j]; }
        x[i] = s / a[i][i];
    }
    x
}
fn lagrange_coeffs(x: &[f64], y: &[f64]) -> Vec<f64> {
    let n = x.len();
    let v: Vec<Vec<f64>> = (0..n).map(|i| (0..n).map(|j| x[i].powi(j as i32)).collect()).collect();
    solve(v, y.to_vec())
}
fn main() {
    let x = [-1.0, 1.0, 2.0, 3.0];
    let y = [-3.0, 1.0, 3.0, 29.0];
    println!("{:?}", lagrange_coeffs(&x, &y));   // [5, -1, -6, 3]
}
